lib/maths/CDataFrameUtils.cc

```diff
@@ -1049,6 +997,263 @@ CDataFrameUtils::metricMicWithColumnDataFrameOnDisk(const CColumnValue& target,
+                 copyObjective, reduceObjective, objective_);
+        return objective_.maxCoeff();
+    };
+
+    auto objectiveGradient = [&](const TDoubleVector& weights) {
+        TDoubleVector probabilities;
+        TDoubleVector scores;
+        auto computeObjectiveAndGradient = core::bindRetrievableState(
+            [=](TDoubleMatrix& state, TRowItr beginRows, TRowItr endRows) mutable {
+                for (auto row = beginRows; row != endRows; ++row) {
+                    if (isMissing((*row)[targetColumn]) == false) {
+                        std::size_t j{static_cast<std::size_t>((*row)[targetColumn])};
+                        probabilities = readPrediction(*row);
+                        CTools::inplaceSoftmax(probabilities);
+                        scores = probabilities.cwiseProduct(weights);
+                        CTools::inplaceSoftmax(scores);
+                        state(j, 0) += (1.0 - scores(j)) / classCounts(j);
+                        state.col(j + 1) +=
+                            scores(j) *
+                            probabilities
+                                .cwiseProduct(scores - TDoubleVector::Unit(numberClasses, j))
+                                .cwiseQuotient(classCounts);
+                    }
+                }
+            },
+            TDoubleMatrix{TDoubleMatrix::Zero(numberClasses, numberClasses + 1)});
+        auto copyObjectiveAndGradient = [](TDoubleMatrix state, TDoubleMatrix& result) {
+            result = std::move(state);
+        };
+        auto reduceObjectiveAndGradient =
+            [&](TDoubleMatrix state, TDoubleMatrix& result) { result += state; };
+
+        TDoubleMatrix objectiveAndGradient;
+        doReduce(frame.readRows(numberThreads, 0, frame.numberRows(),
+                                computeObjectiveAndGradient, &rowMask),
+                 copyObjectiveAndGradient, reduceObjectiveAndGradient, objectiveAndGradient);
+        std::size_t max;
+        objectiveAndGradient.col(0).maxCoeff(&max);
+        return TDoubleVector{objectiveAndGradient.col(max + 1)};
+    };
+
+    TDoubleVector a;
+    TDoubleVector b;
+    std::tie(a, b) = std::move(bounds);
+
+    TMinAccumulator minLoss;
+    TDoubleVector minWeights;
+    TDoubleVector w0{TDoubleVector::Ones(numberClasses)};
+    TDoubleVector interpolate{numberClasses};
+    for (std::size_t i = 0; i < 10; ++i) {
+        CLbfgs<TDoubleVector> lbfgs{5};
+        double loss;
+        std::tie(w0, loss) = lbfgs.minimize(objective, objectiveGradient, std::move(w0));
+        LOG_TRACE(<< "weights = " << w0.transpose() << ", loss = " << loss);
+        if (minLoss.add(loss)) {
+            minWeights = std::move(w0);
+        }
+        for (std::size_t j = 0; j < numberClasses; ++j) {
+            interpolate(j) = CSampling::uniformSample(rng, 0.0, 1.0);
+        }
+        w0 = (a + interpolate.cwiseProduct(b - a)).array().exp();
```

Comment:
It seems to me that this should be at the beginning of the for-loop. Otherwise, you try with w0=(1,1,1..,1) at first and this can be outside of your bounds.

Reply:
I actually intended to do this. My thought was we should always include the best solution in the vicinity of doing nothing, i.e. all weights being equal. Since we use use line search with backtracking we are then *guarantying* that we never do worse than not reweighting at all, which is a nice property since the optimisation objective is complex. WDYT?

Reply:
I reworked initialisation slightly and added a comment.

I still think it is worth trying the one vector for reason outlined. I also now bake in the fact we expect weights to be (roughly) a monotonic decreasing function of class recalls. (This isn't guaranteed because it depends how close probabilities are and what the predicted classes are for the error cases.)

I also reduced the number of restarts because trying out with a wider variety of numbers of classes and range of recalls I didn't see evidence we needed as many restarts after this change. See [this](https://github.com/elastic/ml-cpp/pull/1113/commits/21ba4feb399f86ec251d11ee7f5d840e83f170bf) commit.